src/packet/bwe/trendline_estimator.rs

```diff
@@ -0,0 +1,355 @@
+use std::collections::VecDeque;
+use std::ops::RangeInclusive;
+use std::time::Instant;
+
+use super::{BandwithUsage, InterGroupDelayVariation};
+
+const SMOOTHING_COEF: f64 = 0.9;
+const OVER_USE_THRESHOLD_DEFAULT_MS: f64 = 12.5;
+const OVER_USE_TIME_THRESHOLD_MS: f64 = 10.0;
+const MAX_ADOPT_OFFSET_MS: f64 = 15.0;
+const THRESHOLD_GAIN: f64 = 4.0;
+
+const K_UP: f64 = 0.0087;
+const K_DOWN: f64 = 0.039;
+
+const DELAY_COUNT_RANGE: RangeInclusive<usize> = 60..=1000;
+
+pub(super) struct TrendlineEstimator {
+    /// The window size in packets
+    window_size: usize,
+
+    /// The first instant we saw, used as zero point.
+    zero_time: Option<Instant>,
+
+    /// The history of observed delay variations.
+    history: VecDeque<Timing>,
+
+    /// The total number of observed delay variations.
+    num_delay_variations: usize,
+
+    /// Accumulated delay.
+    accumulated_delay: f64,
+
+    /// Last smoothed delay.
+    smoothed_delay: f64,
+
+    /// The adaptive delay threshold.
+    delay_threshold: f64,
+
+    /// Previous trend
+    previous_trend: f64,
+
+    /// If we are overusing, this contains data about the overuse.
+    overuse: Option<Overuse>,
+
+    /// The last time we updated the adaptive threshold.
+    last_threshold_update: Option<Instant>,
+
+    /// Our current hypothesis about the bandwidth usage.
+    hypothesis: BandwithUsage,
+}
+
+impl TrendlineEstimator {
+    pub(super) fn new(window_size: usize) -> Self {
+        Self {
+            window_size,
+            zero_time: None,
+            history: VecDeque::default(),
```

Comment:
with_capacity(window_size) ?

Reply:
I suppose, I was thinking we should have a type `LimitedHistory` or something that is a FIFO queue with a builtin limit, the pattern is quite common in the codebase